exonum/src/events/noise/mod.rs

```diff
@@ -112,3 +114,174 @@ fn write_handshake_msg(
+            HandshakeParams {
+                max_message_len: 1024,
+                public_key,
+                secret_key,
+            }
+        }
+    }
+
+    #[test]
+    fn test_noise_normal_handshake() {
+        let addr: SocketAddr = "127.0.0.1:45001".parse().unwrap();
+        let addr2 = addr.clone();
+
+        thread::spawn(move || {
+            run_handshake_listener(&addr2).unwrap();
+        });
+
+        // Wait until listener is initialized.
+        thread::sleep(Duration::from_millis(500));
+
+        let res = send_handshake(&addr, HandshakeStep::Default);
+        assert!(res.is_ok());
+    }
+
+    #[test]
+    fn test_noise_handshake_errors() {
+        let addr: SocketAddr = "127.0.0.1:45002".parse().unwrap();
+        let addr2 = addr.clone();
+
+        // Wait until listener is initialized.
+        thread::sleep(Duration::from_millis(500));
+
+        thread::spawn(move || run_handshake_listener(&addr2));
+
+        let res = send_handshake(&addr, HandshakeStep::One(1, EMPTY_MESSAGE));
+        assert!(res.is_err());
+
+        let res = send_handshake(&addr, HandshakeStep::Two(2, EMPTY_MESSAGE));
+        assert!(res.is_err());
+
+        let res = send_handshake(&addr, HandshakeStep::One(1, SMALL_MESSAGE));
+        assert!(res.is_err());
+
+        let res = send_handshake(&addr, HandshakeStep::Two(2, SMALL_MESSAGE));
+        assert!(res.is_err());
+
+        let res = send_handshake(&addr, HandshakeStep::One(1, BIG_MESSAGE));
+        assert!(res.is_err());
+
+        let res = send_handshake(&addr, HandshakeStep::Two(2, BIG_MESSAGE));
+        assert!(res.is_err());
+    }
+
+    fn run_handshake_listener(addr: &SocketAddr) -> Result<(), io::Error> {
+        let mut core = Core::new().unwrap();
+        let handle = core.handle();
+        let params = HandshakeParams::default();
+
+        let fut_stream = TcpListener::bind(addr, &handle).unwrap();
+        let fut = fut_stream
+            .incoming()
+            .for_each(|(stream, _)| {
+                let handshake = NoiseHandshake::listen(&params, stream);
+                let reader = handshake.and_then(|_| Ok(())).map_err(log_error);
+                handle.spawn(reader);
+                Ok(())
+            })
+            .map_err(into_other);
+
+        core.run(fut)
+    }
+
+    fn send_handshake(addr: &SocketAddr, step: HandshakeStep) -> Result<(), io::Error> {
+        let mut core = Core::new().unwrap();
+        let handle = core.handle();
+        let params = HandshakeParams::default();
+
+        let stream = TcpStream::connect(&addr, &handle)
+            .and_then(|sock| match step {
+                HandshakeStep::Default => NoiseHandshake::send(&params, sock),
+                _ => noise_send_handshake_with_error(&params, sock, step),
+            })
+            .map(|_| ())
+            .map_err(into_other);
+
+        core.run(stream)
+    }
+
+    fn noise_send_handshake_with_error(
```

Comment:
Could you briefly explain in comment what aspects this helpers functions adds.

I spent about 15 min to understand test logic.

Reply:
I removed this method after refactoring handshake.